FreeRTOS/FreeRTOS
# Add Full TCP test suite - not using secure sockets

Description
-----------
This PR adds the [Full-TCP test suite](https://github.com/aws/amazon-freertos/blob/master/libraries/abstractions/secure_sockets/test/iot_test_tcp.c#L1080) present in [amazon-FreeRTOS](https://github.com/aws/amazon-freertos) repository after some modifications to the FreeRTOS/FreeRTOS repository. This is a part of a range of additions to the FreeRTOS/FreeRTOS repository to enhance the testing capabilities of this repository. The specific modifications are listed below:
- The Full-TCP test suite uses the Secure-Socket abstraction whereas the tests being added just use FreeRTOS_Sockets since Secure Sockets is not a part of FreeRTOS/FreeRTOS.
- The test names have been changed to reflect the above changes. The names follow this convention: TCP_Sockets_\<*name of function*>_\<*TestCase*>.
- The test case execution order has been changed to make sure that a functionality (e.g. 'FreeRTOS_send') is tested before it is used in some other test-case.


**Note For reviewers**: Important changes are restricted to only 1 file, namely: `test_tcp.c`. (Other files are just additions to the project)
OR
You can try running the demo on your local setup.

Test Steps
-----------
- Clone the Repository with `--recurse-submodules`.
- Go to the 'FreeRTOS-Plus/Test/FreeRTOS-Plus-TCP/Integration/Full-TCP-Suite' directory.
- NOTE: You'll need an active ethernet connection and probably need to switch off any VPN service.
- Open the `Full-TCP-Suite` project.
  - Change the interface to use by modifying the 'configNETWORK_INTERFACE_TO_USE' define.
  - Change the `tcptestECHO_SERVER_ADDR[0-3]` and `tcptestECHO_PORT` in test_tcp.h to reflect the address of your echo server.
  - Save the changes.
- Compile and run the project. It should run 23 tests which should all pass.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
